ext/EvoTreesCUDAExt/fit.jl

```diff
@@ -1,267 +1,330 @@
+            find_split_all! = find_best_split_from_hist_kernel!(backend)
+            find_split_all!(
+                L, view(cache.best_gain_gpu, 1:n_nodes),
+                view(cache.best_bin_gpu, 1:n_nodes),
+                view(cache.best_feat_gpu, 1:n_nodes),
+                cache.h∇, cache.nodes_sum_gpu,
+                view(active_nodes, 1:n_active),
+                cache.js, cache.feattypes_gpu, cache.monotone_constraints_gpu,
+                params.lambda, params.L2, params.min_weight,
+                cache.K, view(cache.sums_temp_gpu, 1:(2*cache.K+1), 1:n_active);
+                ndrange=n_active,
+                workgroupsize=min(256, max(64, n_active)),
+            )
+            KernelAbstractions.synchronize(backend)
+        end
+
+        # Apply splits: create children if gain > threshold, else make leaf
+        apply_splits_kernel!(backend)(
+            cache.tree_split_gpu, cache.tree_cond_bin_gpu, cache.tree_feat_gpu,
+            cache.tree_gain_gpu, cache.nodes_sum_gpu,
+            cache.n_next_gpu, cache.n_next_active_gpu,
+            view(cache.best_gain_gpu, 1:n_nodes),
+            view(cache.best_bin_gpu, 1:n_nodes),
+            view(cache.best_feat_gpu, 1:n_nodes),
+            cache.h∇, active_nodes, cache.feattypes_gpu,
+            depth, params.max_depth, Float32(params.gamma),
+            cache.K;
+            ndrange=max(n_active, 1),
+            workgroupsize=256,
+        )
+        KernelAbstractions.synchronize(backend)
+
+        n_active = Array(cache.n_next_active_gpu)[1]
+        if n_active > 0
+            copyto!(view(cache.anodes_gpu, 1:n_active), view(cache.n_next_gpu, 1:n_active))
+        end
+
+        # Update observation→node assignments for next level
+        # Finalize nidx at max depth to ensure MAE/Quantile leaf predictions use correct observation sets
+        if depth <= params.max_depth && n_active > 0
+            update_nodes_idx_kernel!(backend)(
+                cache.nidx, is, cache.x_bin, cache.tree_feat_gpu,
+                cache.tree_cond_bin_gpu, cache.feattypes_gpu;
+                ndrange=length(is),
+                workgroupsize=256,
+            )
+            KernelAbstractions.synchronize(backend)
+        end
+    end
+
+    # Copy tree to CPU and compute leaf predictions
+    copyto!(tree.split, cache.tree_split_gpu)
+    copyto!(tree.feat, cache.tree_feat_gpu)
+    copyto!(tree.cond_bin, cache.tree_cond_bin_gpu)
+    copyto!(tree.gain, cache.tree_gain_gpu)
+    copyto!(tree.w, view(cache.nodes_sum_gpu, size(cache.nodes_sum_gpu, 1), 1:length(tree.w))) # FIXME: nodes_sum_gpu[k,:] should match tree.w size
+
+    leaf_nodes = findall(!, tree.split)
+
+    # MAE/Quantile: use CPU-based leaf prediction with exact observation indices
+    # Build leaf membership from nidx and compute predictions using actual sample quantiles
+    # instead of histogram-accumulated sums to avoid numerical errors.
+    if L <: Union{EvoTrees.MAE,EvoTrees.Quantile}
```

Comment:
Special loss treatment, where individual observations is needed, is specific  to `Quantile`. 
`MAE` should work following the generic approacah (even if it's not the regular gradient update rule, it only depends on aggregate metrics provided by the histograms).
See: https://github.com/Evovest/EvoTrees.jl/blob/d81c6033ab56264e02606264d008d84bf5e481e0/src/fit.jl#L132-L135